### PR TITLE
Implement generic Storage protocol + Supabase storage implementation

### PR DESCRIPTION
**Pull request by jdc123**
Supabase storage implementation built on top of generic Storage protocol

### CODE REVIEW

File: src/fixpoint/storage/protocol.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""Protocol for the storage"""
+
+from typing import Any, Optional, Protocol, Dict
+
+
+class SupportsStorage(Protocol):
```
Comment (dbmikus) [inline src/fixpoint/storage/protocol.py:6]: In here, it is unclear what `resource`, `data`, and `model` mean at first glance. To me, `data` and `resource` mean the same thing.

I vote for these changes:

1. `resource` becomes `resource_type`
2. `data` stays as is
3. I think `model` is fine, but maybe just add a comment describing what its purpose is.

File: src/fixpoint/storage/supabase.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        try:
+            self._client: Client = create_client(url, key)
+        except Exception as e:
+            raise ConnectionError(f"Failed to connect to Supabase: {e}") from e
+
+    def _query_table(self, resource: Optional[str] = None) -> Any:
+        lookup_table = resource or self._default_table
+        if lookup_table is None:
+            raise ValueError(
+                "Either table must be provided or default_table must be set"
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:22]: This comment is wrong. The argument is `resource` (or as I propose, `resource_type`)

File: src/fixpoint/storage/supabase.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    def update(
+        self,
+        data: Any,
+        resource: Optional[str] = None,
+        model: Optional[Any] = None,
+    ) -> Any:
+        """Update items in storage (uses upsert)"""
+        try:
+            query = self._query_table(resource)
+            result = query.upsert(data).execute()
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:86]: How does it know which row of the table it is updating? Sorry, I am not familiar with this library.

File: src/fixpoint/utils/db/postgres.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+"""Utility functions for the database"""
+
+from typing import Any
+from psycopg import Connection
+from psycopg.rows import TupleRow
+
+
+def execute_sql(connection: Connection[TupleRow], sql: str, params: Any = None) -> None:
```
Comment (dbmikus) [inline src/fixpoint/utils/db/postgres.py:8]: Are others supposed to use this? You are allowing for SQL injection here

File: src/fixpoint/utils/db/postgres.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+"""Utility functions for the database"""
```
Comment (dbmikus) [inline src/fixpoint/utils/db/postgres.py:1]: Should this be a public module or a private module?

File: src/fixpoint/utils/env.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+"""Utility functions for the environment"""
+
+import os
+
+
+def get_env_value(key: str) -> str:
+    """Get the value of an environment variable"""
+    value = os.environ.get(key)
+    if value is None:
+        raise EnvironmentError(f"Invalid {key}. Please specify.")
```
Comment (dbmikus) [inline src/fixpoint/utils/env.py:10]: Error message seems incomplete

File: src/fixpoint/storage/supabase.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
+"""Supabase storage"""
+
+from typing import Any, Optional
+from supabase import create_client, Client
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:4]: I'd add:

```
from postgrest import SyncRequestBuilder
```

Then you can use that as the return type from `_query_table`

File: src/fixpoint/storage/supabase.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+from .protocol import SupportsStorage
+
+
+class SupabaseStorage(SupportsStorage):
+    """Supabase storage"""
+
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:10]: nit: I think it is good practice to define the class fields here, so:

```
_default_table: Optional[str]
_client: Client
```

Makes it easy to see what fields we have available to work with.

File: src/fixpoint/storage/supabase.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    def __init__(self, url: str, key: str, default_table: Optional[str] = None):
+        self._default_table = default_table
+        try:
+            self._client: Client = create_client(url, key)
+        except Exception as e:
+            raise ConnectionError(f"Failed to connect to Supabase: {e}") from e
+
+    def _query_table(self, resource: Optional[str] = None) -> Any:
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:18]: Change the return type to `SyncRequestBuilder`

File: src/fixpoint/storage/supabase.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                query = query.eq(k, v)
+            result = query.execute()
+            if model:
+                return [model(**item) for item in result.data]
+            return result.data
+        except Exception as e:
+            raise RuntimeError(f"Failed to fetch data: {e}") from e
+
+    def insert(
+        self, data: Any, resource: Optional[str] = None, model: Optional[Any] = None
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:65]: If you fix the return type of `def _query_table(...)`, then you can see the data types when you call `query.insert`. It appears that `data: Union[List, Dict]`, so we should fix that. Also, update `model: Optional[Type[BaseModel]]`.

File: src/fixpoint/storage/supabase.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                return [model(**item) for item in result.data]
+            return result.data
+        except Exception as e:
+            raise RuntimeError(f"Failed to insert data: {e}") from e
+
+    def update(
+        self,
+        data: Any,
+        resource: Optional[str] = None,
+        model: Optional[Any] = None,
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:81]: You can fix the `data` and `model` type definitions here, too

File: src/fixpoint/storage/supabase.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            return result.data
+        except Exception as e:
+            raise RuntimeError(f"Failed to update data: {e}") from e
+
+    def delete(
+        self, filter_criteria: dict[str, Any], resource: Optional[str] = None
+    ) -> Any:
+        """Delete items from storage that match the keys"""
+        try:
+            query = self._query_table(resource).delete()
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:98]: It looks like the default behavior here is to delete everything from the table? The other alternative is to require at least one `filter_critera` entry. I think I like that behavior less, so I'd leave it as you have it.

**Review by dbmikus — COMMENTED** [review]
I think the generic storage interfaces should be private to the package. Because they operate at such an abstract level (they could conceivably be in any Python package), there is no point exposing them as part of our API surface to our users. It will clutter the API surface, and it will not bring the users value.

We don't need to do this right now, necessarily. I think we should do a sweep over the API surface when we are about to share the package and decide what is a public part of our API and what is private.

File: src/fixpoint/storage/supabase.py (hunk 9 above)
Comment (dbmikus) [reply src/fixpoint/storage/supabase.py:18]: BTW, you can do

```
from postgrest import SyncRequestBuilder
```

File: src/fixpoint/storage/supabase.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        self._table = table
+        self._order_key = order_key
+        self._id_column = id_column
+        self._value_type = value_type
+        try:
+            self._client: Client = create_client(url, key)
+        except Exception as e:
+            raise ConnectionError(f"Failed to connect to Supabase: {e}") from e
+
+    def _query_table(self) -> Any:
```
Comment (dbmikus) [inline src/fixpoint/storage/supabase.py:31]: I'd add:

```
from postgrest import SyncRequestBuilder
```

Then you can use that as the return type from `_query_table`. And then all callers of `_query_table` will have useful type information.

File: src/fixpoint/utils/db/postgres.py (hunk 5 above)
Comment (dbmikus) [reply src/fixpoint/utils/db/postgres.py:1]: Bump on this. Does this module exist for testing purposes? I believe so, and for that reason think we should put it inside the root `tests/` folder.

File: src/fixpoint/utils/db/postgres.py (hunk 4 above)
Comment (dbmikus) [reply src/fixpoint/utils/db/postgres.py:8]: This allows SQL injection. I would want to make sure this is never exposed through any user APIs.

If this file goes in the `tests/` folder, then I don't care about this issue. If it stays here, then I think we need to name this something like `UNSAFE_execute_sql`, similar to what [React does](https://react.dev/reference/react/Component#unsafe_componentwillmount).

File: src/fixpoint/utils/env.py (hunk 6 above)
Comment (dbmikus) [reply src/fixpoint/utils/env.py:10]: Bump

**Review by dbmikus — COMMENTED** [review]
Overall, looks good! I have some unresolved comments

File: src/fixpoint/utils/db/postgres.py (hunk 5 above)
Comment (jdc123) [reply src/fixpoint/utils/db/postgres.py:1]: It was to be used for testing primarily, but thought that maybe we could utilize it somewhere in our implementation code at some point. I can move it into tests.

**Pull request merged**